tests/backup/backup_helper.go

```diff
@@ -1629,6 +1667,310 @@ func restoreSuccessWithReplacePolicy(restoreName string, orgID string, retryDura
+
+		// looping over the list of volumes that PX-Backup says it restored, to run some checks
+		for _, restoredVolInfo := range apparentlyRestoredVolumes {
+			if namespaceMappings[restoredVolInfo.SourceNamespace] == expectedRestoredAppContextNamespace {
+				if restoredVolInfo.Status.Status != api.RestoreInfo_StatusInfo_Success /*Can this also be partialsuccess?*/ {
+					err := fmt.Errorf("in restore [%s], the status of the restored volume [%s] was not Success. It was [%s] with reason [%s]", restoreName, restoredVolInfo.RestoreVolume, restoredVolInfo.Status.Status, restoredVolInfo.Status.Reason)
+					errors = append(errors, err)
+					continue
+				}
+
+				var actualVol *volume.Volume
+				var ok bool
+				if actualVol, ok = actualVolumeMap[restoredVolInfo.RestoreVolume]; !ok {
+					err := fmt.Errorf("in restore [%s], said restored volume [%s] cannot be found in the actual cluster [%s]", restoreName, restoredVolInfo.RestoreVolume, expectedRestoredAppContextNamespace)
+					errors = append(errors, err)
+					continue
+				}
+
+				if actualVol.Name != restoredVolInfo.Pvc {
+					err := fmt.Errorf("in restore [%s], for restored volume [%s], PVC used is given as [%s], but actual PVC used was found to be [%s]", restoreName, restoredVolInfo.RestoreVolume, restoredVolInfo.Pvc, actualVol.Name)
+					errors = append(errors, err)
+				}
+
+				var expectedVolumeDriver string
+				if os.Getenv("BACKUP_TYPE") == "Generic" {
+					expectedVolumeDriver = "kdmp"
+				} else {
+					expectedVolumeDriver = Inst().V.String()
+				}
+
+				if restoredVolInfo.DriverName != expectedVolumeDriver {
+					err := fmt.Errorf("in restore [%s], for restored volume [%s], volume driver actually used is given as [%s], but expected is [%s]", restoreName, restoredVolInfo.RestoreVolume, restoredVolInfo.DriverName, expectedVolumeDriver)
+					errors = append(errors, err)
+				}
+			}
+		}
+
+		// VALIDATE APPLICATIONS
+		log.InfoD("Validate applications in restored namespace [%s] due to restore [%s]", expectedRestoredAppContextNamespace, restoreName)
+		errorChan := make(chan error, errorChannelSize)
+		ValidateContext(expectedRestoredAppContext, &errorChan)
+		for err := range errorChan {
+			errors = append(errors, err)
+		}
+	}
+
+	errStrings := make([]string, 0)
+	for _, err := range errors {
+		if err != nil {
+			errStrings = append(errStrings, err.Error())
+		}
+	}
+
+	if len(errStrings) > 0 {
+		return fmt.Errorf("ValidateRestore Errors: {%s}", strings.Join(errStrings, "}\n{"))
+	} else {
+		return nil
+	}
+}
+
+// TransformAppContextWithMappings clones an appContext and transforms it according to the maps provided. When `forRestore` is true, VolumeSnapshots and StorageClasses are ignored. To be used after switching k8s context (cluster) which has the namespace
+func TransformAppContextWithMappings(appContext *scheduler.Context, namespaceMapping map[string]string, storageClassMapping map[string]string, forRestore bool) (*scheduler.Context, error) {
```

Comment:
How about using DeepCopy of scheduler.Context?

Reply:
I don't think that function works properly. It's missing some fields.

Reply:
Yes, but modifying the context will alter the original value since it is a pointer to the context.

Reply:
no, changed function name to be more clear. It clones and transforms